File: langwatch_nlp/langwatch_nlp/studio/dspy/http_node.py
"""
HTTP node module for dspy workflows.

Wraps the HTTP node executor as a dspy.Module for use in workflow execution.
"""

import asyncio
from typing import Any, Awaitable, Callable, Dict, Optional, TypeVar

import dspy

from langwatch_nlp.studio.execute.http_node import (
    HttpNodeConfig,
    HttpAuthConfig,
    execute_http_node,
)

T = TypeVar("T")


def run_async_safely(coro: Callable[..., Awaitable[T]], *args: Any, **kwargs: Any) -> T:
    """Run an async function safely from sync context.

    Handles both cases:
    - When called from within an existing event loop (uses asyncer.syncify)
    - When called from sync context without a running loop (uses asyncio.run)

    Args:
        coro: The async function to call
        *args: Positional arguments to pass to the function
        **kwargs: Keyword arguments to pass to the function

    Returns:
        The result of the async function
    """
    try:
        asyncio.get_running_loop()
        # If we're already in an async context, use asyncer to bridge
        import asyncer

        return asyncer.syncify(coro)(*args, **kwargs)
    except RuntimeError:
        # No running loop, create one
        return asyncio.run(coro(*args, **kwargs))


class HttpNode(dspy.Module):
    """DSPy module that executes HTTP requests."""

    def __init__(
        self,
        url: str,
        method: str = "POST",
        body_template: Optional[str] = None,
        output_path: Optional[str] = None,
        auth_type: Optional[str] = None,
        auth_token: Optional[str] = None,
        auth_header: Optional[str] = None,
        auth_value: Optional[str] = None,
        auth_username: Optional[str] = None,
        auth_password: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        timeout_ms: Optional[int] = None,
    ):
        super().__init__()

        # Build auth config if provided
        auth: Optional[HttpAuthConfig] = None
        if auth_type:
            auth = HttpAuthConfig(
                type=auth_type,  # type: ignore
                token=auth_token,
                header=auth_header,
                value=auth_value,
                username=auth_username,
                password=auth_password,
            )

        self.config = HttpNodeConfig(
            url=url,
            method=method,  # type: ignore
            body_template=body_template,
            output_path=output_path,
            auth=auth,
            headers=headers,
            timeout_ms=timeout_ms,
        )

    def forward(self, **kwargs: Any) -> Any:
        """Execute the HTTP request with the given inputs.

        Args:
            **kwargs: Input values for template interpolation

        Returns:
            The extracted output from the HTTP response

        Raises:
            Exception: If the HTTP request fails
        """
        result = run_async_safely(execute_http_node, self.config, kwargs)

        if not result.success:
            raise Exception(result.error or "HTTP request failed")

        return {"output": result.output}

File: packages/features/agent/contract/src/http-node.ts
import type { Field } from "./fields";
import type { HttpAuth, HttpHeader, HttpMethod } from "./config/http";

export type HttpCallConfig = {
  url: string;
  method?: HttpMethod;
  headers?: HttpHeader[];
  auth?: HttpAuth;
  bodyTemplate?: string;
  outputPath?: string;
  timeoutMs?: number;
};

export function buildHttpNodeParameters(config: HttpCallConfig): Field[] {
  const parameters: Field[] = [
    { identifier: "url", type: "str", value: config.url },
    { identifier: "method", type: "str", value: config.method ?? "POST" },
  ];
  if (config.bodyTemplate)
    parameters.push({
      identifier: "body_template",
      type: "str",
      value: config.bodyTemplate,
    });
  if (config.outputPath)
    parameters.push({
      identifier: "output_path",
      type: "str",
      value: config.outputPath,
    });
  const headers = Object.fromEntries(
    (config.headers ?? [])
      .map(({ key, value }) => [key.trim(), value] as const)
      .filter(([key]) => key.length > 0),
  );
  if (Object.keys(headers).length > 0)
    parameters.push({ identifier: "headers", type: "dict", value: headers });
  if (config.timeoutMs)
    parameters.push({
      identifier: "timeout_ms",
      type: "int",
      value: config.timeoutMs,
    });
  if (config.auth && config.auth.type !== "none") {
    parameters.push({
      identifier: "auth_type",
      type: "str",
      value: config.auth.type,
    });
    if (config.auth.type === "bearer")
      parameters.push({
        identifier: "auth_token",
        type: "str",
        value: config.auth.token,
      });
    if (config.auth.type === "api_key") {
      parameters.push({
        identifier: "auth_header",
        type: "str",
        value: config.auth.header,
      });
      parameters.push({
        identifier: "auth_value",
        type: "str",
        value: config.auth.value,
      });
    }
    if (config.auth.type === "basic") {
      parameters.push({
        identifier: "auth_username",
        type: "str",
        value: config.auth.username,
      });
      parameters.push({
        identifier: "auth_password",
        type: "str",
        value: config.auth.password,
      });
    }
  }
  return parameters;
}
